### PR TITLE
Dashboard: Store original JSON in DashboardModel

### PR DESCRIPTION
**Pull request by Clarity-89**
**What is this feature?**

Save the original dashboard JSON inside the DashboardModel instead of making an extra API request to fetch it when saving. 

**Why do we need this feature?**

Removes an extra API request for fetching dashboard JSON when checking if there were any changes to the dashboard on save.

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/dashboard/state/reducers.ts [hunk 1]
```diff
@@ -30,9 +31,13 @@ const dashboardSlice = createSlice({
...
     },
-    dashboardInitCompleted: (state, action: PayloadAction<DashboardModel>) => {
-      state.getModel = () => action.payload;
+    dashboardInitCompleted: (
+      state,
+      action: PayloadAction<{ dashboard: DashboardModel; original: DashboardDataDTO | null }>
+    ) => {
+      state.getModel = () => action.payload.dashboard;
       state.initPhase = DashboardInitPhase.Completed;
+      state.getOriginal = () => action.payload.original;
```
Comment (Clarity-89) [inline public/app/features/dashboard/state/reducers.ts:40]: Tried saving the JSON directly without a wrapper function but get this error:
Screenshot 2023-08-28 at 8 53 57

Comment (torkelo) [reply public/app/features/dashboard/state/reducers.ts:40]: ah, yes. I think we modify the object to remove all the properties that should not be included in the diff

**Review by dprokop — COMMENTED** [review]
Looks good. 

Thought - do we need to save in in Store? For the sake of simplicity, maybe just having it as a private property with a getter on DashboardModel would be sufficient, given the original dashboard is already passed in there via constructor, and dashboard model is available downstream the components on Dashboard page.

**Comment by Clarity-89** [comment]
> Thought - do we need to save in in Store? For the sake of simplicity, maybe just having it as a private property with a getter on DashboardModel would be sufficient, given the original dashboard is already passed in there via constructor, and dashboard model is available downstream the components on Dashboard page.

That's a good point and should make the code a bit simpler.

File: public/app/features/dashboard/state/initDashboard.ts [hunk 2]
```diff
@@ -283,6 +283,7 @@ export function initDashboard(args: InitDashboardArgs): ThunkResult<void> {
     trackDashboardLoaded(dashboard, versionBeforeMigration);
 
     // yay we are done
+    dashboard.setOriginalDashboard(dashDTO.dashboard);
```
Comment (dprokop) [inline public/app/features/dashboard/state/initDashboard.ts:286]: I think this setter is unnecessary. This model is passed to DashboardModel constructor on L193 and can be used to store the private property.

Comment (Clarity-89) [reply public/app/features/dashboard/state/initDashboard.ts]: Good catch!

**Review by dprokop — APPROVED** [review]

**Pull request merged**